msal/src/Microsoft.Identity.Client/Internal/Requests/DeviceCodeRequest.cs

```diff
@@ -0,0 +1,125 @@
+    // * do we have lab accounts/environments setup for device auth?
+    // * do we need to set SaveToCache in the constructor?
+    internal class DeviceCodeRequest : RequestBase
+    {
+        private readonly Action<DeviceCodeResult> _deviceCodeResultCallback;
+        private DeviceCodeResult _deviceCodeResult;
+
+        public DeviceCodeRequest(AuthenticationRequestParameters authenticationRequestParameters, Action<DeviceCodeResult> deviceCodeResultCallback) 
+            : base(authenticationRequestParameters)
+        {
+            _deviceCodeResultCallback = deviceCodeResultCallback;
+            
+            LoadFromCache = false;  // no cache lookup for token
+            SupportADFS = false;
+        }
+
+        internal override async Task PreTokenRequestAsync()
+        {
+            await base.PreTokenRequestAsync().ConfigureAwait(false);
+
+            OAuth2Client client = new OAuth2Client();
+
+            var deviceCodeScopes = new SortedSet<string>();
+            deviceCodeScopes.UnionWith(AuthenticationRequestParameters.Scope);
+            deviceCodeScopes.Add("offline_access");
+            deviceCodeScopes.Add("profile");
+            deviceCodeScopes.Add("openid");
+
+            client.AddBodyParameter(OAuth2Parameter.ClientId, AuthenticationRequestParameters.ClientId);
+            client.AddBodyParameter(OAuth2Parameter.Scope, deviceCodeScopes.AsSingleString());
+
+            // todo: THIS IS A MAJOR HACK.  Work with Shiung/Henrik on proper way to determine the device code endpoint
+            string deviceCodeEndpoint = AuthenticationRequestParameters.Authority.TokenEndpoint
+                .Replace("token", "devicecode")
+                .Replace("common", "organizations");
+
+            AuthenticationRequestParameters.Authority.TokenEndpoint = AuthenticationRequestParameters.Authority.TokenEndpoint.Replace("common", "organizations");
+
+            DeviceCodeResponse response = await client.ExecuteRequestAsync<DeviceCodeResponse>(
+                client.CreateFullEndpointUri(new Uri(deviceCodeEndpoint)), 
+                HttpMethod.Post,
+                AuthenticationRequestParameters.RequestContext).ConfigureAwait(false);
+
+            // todo: fix resource/scopes here for how we want to invoke this...
+            _deviceCodeResult = response.GetResult(AuthenticationRequestParameters.ClientId, "todo resource");
+            _deviceCodeResultCallback(_deviceCodeResult);
+        }
+
+        protected override async Task SendTokenRequestAsync()
+        {
+            TimeSpan timeRemaining = _deviceCodeResult.ExpiresOn - DateTimeOffset.UtcNow;
+
+            while (timeRemaining.TotalSeconds > 0.0)
+            {
+                try
+                {
+                    await base.SendTokenRequestAsync().ConfigureAwait(false);
+                    return;
+                }
+                catch (MsalServiceException ex)

```

Comment:
>catch [](start = 16, length = 5)

we may have to also handle the "429" here, unless we handle that either at lower level or at a higher layer...